EIP-0023 Oracle pool 2.0 (#41)
ergoplatform/eips: eip-0023/eip-0023.md

|Oracle | 15 | Oracle token <br /> Reward tokens | R4: Public key (GroupElement) <br /> R5: Epoch counter of pool box (Int) <br /> R6: Published rate (Long) | Publish data-point <br /> Accumulate reward tokens | Publish data-point, <br /> Refresh pool, <br /> Transfer oracle token, <br /> Extract reward tokens|
|Update | 1 | Update-NFT | | Updating pool box | Update pool box |
|Ballot | 15 | Ballot token | R4: Public key (GroupElement) <br /> R5: Update box creation height (Int) <br /> R6: New pool box hash (Coll[Byte]) | Voting for updating pool box | Vote for update <br /> Update pool box <br /> Transfer ballot token |

Before going into the transactions in the protocol, we first present the contracts for each of the above boxes.

## Contracts with base-64-encoded hash of ergo-tree bytes

- [Pool Contract](contracts/pool_contract.es) `8cJi+FGGU32jXyO8M2LeyWSWlerdcb1zxBWeZtyy7Y8=`
- [Refresh Contract](contracts/refresh_contract.es) `cs5c5QEirstI4ZlTyrbTjlPwWYHRW+QsedtpyOSBnH4=`
- [Oracle Contract](contracts/oracle_contract.es) `fhOYLO3s+NJCqTQDWUz0E+ffy2T1VG7ZnhSFs0RP948=`
- [Ballot Contract](contracts/ballot_contract.es) `2DnK+72bh+TxviNk8XfuYzLKtuF5jnqUJOzimt30NvI=`

> - [Update Contract](contracts/update_contract.es) `0wFmk/1TNpgTsbzpWND3WLPbwQdD8E+TWDzZLaYv3nE=`

Reviewer: ```suggestion
- [Update Contract](contracts/update_contract.es) `3aIVTP5tRgCZHxaT3ZFw3XubRV5DJi0rKeo9bKVHlVw=`
```
Generated with https://scastie.scala-lang.org/BxJFCRDcSCeiwf9ZgKdymQ
using the latest update contract https://github.com/ergoplatform/eips/blob/1757e0060d6dafa6bd9bcb2339ab5484dfc767a7/eip-0023/contracts/update_contract.es with pool box preservation requirement removed.
Corresponding PR in oracle-core is https://github.com/ergoplatform/oracle-core/pull/178

> Use this Scastie playground to calculate the above hashes - [https://scastie.scala-lang.org/hnTEm2lJQPG1wRYCqPz8LQ](https://scastie.scala-lang.org/hnTEm2lJQPG1wRYCqPz8LQ)

Reviewer: formatting

```suggestion

Use this Scastie playground to calculate the above hashes - [https://scastie.scala-lang.org/hnTEm2lJQPG1wRYCqPz8LQ](https://scastie.scala-lang.org/hnTEm2lJQPG1wRYCqPz8LQ)
```

   - It should store the rate to publish in R6.
   - It should ensure reward tokens are preserved.

### Extract reward tokens

| Index | Input | Output |
|---|---|---|
| 0 | Oracle | Oracle |
| 1 |             | Box with freed reward tokens |

This allows an oracle to extract tokens obtained from publishing data point to redeem them elsewhere.

The transaction is similar to the [publish data-point](#publish-data-point) transaction, except that Step 3 is modified as follows.

3. The following rules (not enforced by the smart contract) to be followed.
   - It should store the same group element in R4.
   - It should keep at least 1 reward token.
   - It should store the balance reward tokens in some other box.

### Transfer oracle token

| Index | Input | Output |
|---|---|---|
| 0 | Oracle | Oracle |  

This is used to transfer the ownership of an oracle token to another public key.

The transaction is similar to the [publish data-point](#publish-data-point) transaction, except that Step 3 is modified as follows.

3. The following rules (not enforced by the smart contract) to be followed.
   - It should store the new owner's group element in R4.

>    - It should keep at least 1 reward token.
>    - It should store the balance reward tokens, if any, in some other box.

Reviewer: Following our discussion in https://github.com/ergoplatform/eips/pull/41#discussion_r870469418
```suggestion
   - It should check that there is only 1 reward token is in the oracle box and send it to the new owner along with oracle token.
```
